lib/internal/encoding.js

```diff
@@ -0,0 +1,359 @@
+  ['x-euc-jp', 'euc-jp'],
+  ['csiso2022jp', 'iso-2022-jp'],
+  ['iso-2022-jp', 'iso-2022-jp'],
+  ['csshiftjis', 'shift_jis'],
+  ['ms932', 'shift_jis'],
+  ['ms_kanji', 'shift_jis'],
+  ['shift-jis', 'shift_jis'],
+  ['shift_jis', 'shift_jis'],
+  ['sjis', 'shift_jis'],
+  ['windows-31j', 'shift_jis'],
+  ['x-sjis', 'shift_jis'],
+  ['cseuckr', 'euc-kr'],
+  ['csksc56011987', 'euc-kr'],
+  ['euc-kr', 'euc-kr'],
+  ['iso-ir-149', 'euc-kr'],
+  ['korean', 'euc-kr'],
+  ['ks_c_5601-1987', 'euc-kr'],
+  ['ks_c_5601-1989', 'euc-kr'],
+  ['ksc5601', 'euc-kr'],
+  ['ksc_5601', 'euc-kr'],
+  ['windows-949', 'euc-kr'],
+  ['utf-16be', 'utf-16be'],
+  ['utf-16le', 'utf-16le'],
+  ['utf-16', 'utf-16le'],
+  // Unsupported standard mappings
+  // ['csiso2022kr', 'replacement'],
+  // ['hz-gb-2312', 'replacement'],
+  // ['iso-2022-cn', 'replacement'],
+  // ['iso-2022-cn-ext', 'replacement'],
+  // ['iso-2022-kr', 'replacement'],
+  // ['x-user-defined', 'x-user-defined'],
+  // ['iso-8859-16', 'iso-8859-16'],
+]);
+
+function normalizeEncoding(encoding) {
+  if (encoding === undefined)
+    return 'utf-8';
+  let enc = encodings.get(encoding);
+  if (enc !== undefined) return enc;
+  enc = encodings.get(encoding.toLowerCase());
+  if (enc !== undefined) return enc;
+}
+
+function hasTextDecoder(encoding = 'utf-8') {
+  if (typeof encoding !== 'string')
+    throw new errors.Error('ERR_INVALID_ARG_TYPE', 'encoding', 'string');
+  return hasConverter(normalizeEncoding(encoding));
+}
+
+class TextDecoder {
+  constructor(encoding = 'utf-8', options = {}) {
+    if (typeof encoding !== 'string')
+      throw new errors.Error('ERR_INVALID_ARG_TYPE', 'encoding', 'string');
+    if (options !== undefined && typeof options !== 'object')
+      throw new errors.Error('ERR_INVALID_ARG_TYPE', 'options', 'object');
+
+    const fatal = options !== null &&
+                  options.fatal === true;
+
+    const ignoreBOM = options !== null &&
+                      options.ignoreBOM === true;
+
+    const enc = normalizeEncoding(encoding);
+    if (enc === undefined || enc === 'replacement')
+      throw new errors.RangeError('ERR_ENCODING_NOT_SUPPORTED', encoding);
+
+    var flags = 0;
+    flags |= fatal ? CONVERTER_FLAGS_FATAL : 0;
+    flags |= ignoreBOM ? CONVERTER_FLAGS_IGNORE_BOM : 0;
+
+    const handle = getConverter(enc, flags);
+    if (handle === undefined)
+      throw new errors.Error('ERR_ENCODING_NOT_SUPPORTED', encoding);
+
+    Object.defineProperties(this, {
+      [kHandle]: {
+        configurable: false,
+        enumerable: false,
+        value: handle
+      },
+      encoding: {
+        configurable: false,
+        enumerable: true,
+        value: enc
+      },
+      fatal: {
+        configurable: false,
+        enumerable: true,
+        value: fatal
+      },
+      ignoreBOM: {
+        configurable: false,
+        enumerable: true,
+        value: ignoreBOM
+      }
+    });
+  }
+
+  decode(input, options = {}) {
```

Comment:
For strict IDL spec compliance, `decode.length` needs to be 0, as `input` is annotated with `optional`. `decode(input = undefined, options = {})` should do the job. `decode` should also be enumerable, which means that you might need to manually do `TextDecoder.prototype.decode = function` or use a trick like [this](https://github.com/nodejs/node/blob/d71718db6aa4feb8dc10edbad1134472468e971a/lib/internal/url.js#L780-L781).